README.md

```diff
@@ -1,16 +1,1261 @@
+
+``` json
+// CID
+{"/": "Qmf412jQZiuVUtdgnB36FXFX7xg5V6KEbSJ4dpQuhkLyfD"}
+
+// Bytes (e.g. a signature)
+{"/": {"bytes": "s0m3Byte5"}}
+```
+
+This format help disambiguate type information in generic DAG-JSON tooling. However, your presentation need not be in this specific format, as long as it can be converted to and from this cleanly. As it is used for the signature format, DAG-CBOR is RECOMMENDED.
+
+## 1.4 Signatures
+
+All payloads described in this spec MUST be signed with a [Varsig](https://github.com/ChainAgnostic/varsig/).
+
+# 2 High-Level Concepts
+
+## 2.1 Roles
+
+Task adds two new roles to UCAN: invoker and executor. The existing UCAN delegator and delegate principals MUST persist to the invocation.
+
+| UCAN Field | Delegation                             | Task                      |
+|------------|----------------------------------------|---------------------------------|
+| `iss`      | Delegator: transfer authority (active) | Invoker: request task (active)  |
+| `aud`      | Delegate: gain authority (passive)     | Executor: perform task (active) |
+
+### 2.1.1 Invoker
+
+The invoker signals to the executor that a task associated with a UCAN SHOULD be performed.
+
+The invoker MUST be the UCAN delegator. Their DID MUST be authenticated in the `iss` field of the contained UCAN.
+
+### 2.1.2 Executor
+
+The executor is directed to perform some task described in the UCAN by the invoker.
+
+The executor MUST be the UCAN delegate. Their DID MUST be set the in `aud` field of the contained UCAN.
+
+## 2.2 Components 
+
+![](./diagrams/concepts.svg)
+
+### 2.2.1 Closure
+
+A [Closure](#3-closure) is like a deferred function application: a request to perform some action on a resource with specific inputs.
+
+### 2.2.2 Task
+
+A [Task](#4-task) extends a Closure with additional metadata that is not used to describe the meaning of the computation or effect to be run.
+
+### 2.2.3 Batch
+
+A [Batch](#5-batch) is a way of requesting more than one action at once.
+
+### 2.2.4 Invocation
+
+An [Invocation](#6-invocation) is the cryptographically signed container for a Batch. This is the step that "forces" the "deferred" Closure.
+
+### 2.2.5 Pointers
+
+An [Invocation Pointer](#7-pointer) identifies a specific invocation. An Invoked Task Pointer points to a unique Task inside an Invocation.
+
+### 2.2.6 Result
+
+A [Result](#8-result) is the output of a Closure.
+
+### 2.2.7 Receipt
+
+A [Receipt](#9-receipt) describes the output of an invocation, referenced by its Invocation Pointer.
+
+### 2.2.8 Promise
+
+A [promise](#10-promise) is a reference to the receipt of an action that has yet to return a receipt.
+
+## 2.3 IPLD Schema
+
+``` ipldsch
+type Closure struct {
+  with   URI
+  do     Ability
+  inputs Any
+}
+
+type Task struct {
+  with   URI
```

Comment:
I think my main question after a first skim is: what's the difference between "with" and "inputs"?  Why isn't "with" just some kind of member or specially-named map key within "inputs"?

By contrast, I get why "meta" is broken out (tldr: memoization / what is useful to hash).  But that doesn't seem to apply to "with"?

Reply:
@warpfork You can see all of these as "inputs" to the computation, sure. The difference is that `with` is the thing being called.

```js
with.action(inputs)
```

You're right that we can rearrange this to look like:

```js
```

Reply:
Also: my understanding is that IPLD Schema doens't have a concept of subtyping, though you can emulate it manually. You're probably the right person to ask: is that the case?